### ISSUE TITLE
[AMD][kernel][Tier 5] Port sol-execbench jamba_attn_proj (cuTile → FlyDSL) to FlyDSL

### ISSUE THREAD

**Issue by sunxxuns** [amdpilot-org/FlyDSL#41]
# FlyDSL implementation: port sol-execbench's jamba_attn_proj problem (cuTile variant) to FlyDSL

Cross-references:
- `amdpilot-org/SOL-ExecBench` — AMD fork of `nvidia/sol-execbench` (the GPU kernel eval framework that scores AI-generated kernel solutions for correctness + performance against a PyTorch reference). Currently NVIDIA-only: CUDA / Triton / CUTLASS / CuTe DSL / cuTile.
- `examples/cutile/jamba_attn_proj/` in that fork — the cuTile variant of the fused attention output projection with residual. The `definition.json` names it `ai21labs_AI21-Jamba-Reasoning-3B_attention_output_projection_with_residual`; the reference is captured from AI21's Jamba-Reasoning-3B attention o_proj.
- Sibling FlyDSL issue: **#39 (cute_dsl/jamba_attn_proj)** — being filed in parallel by a sibling agent. The `cute_dsl/jamba_attn_proj` variant solves the **same** definition (same PyTorch reference, same workloads, same thresholds) but the NVIDIA solution is authored in CUTLASS CuTe DSL (`cutlass.cute`) instead of cuTile (`cuda.tile`). **The two issues should be tracked together** — see §"Why this issue exists" below.
- Prior FlyDSL issue: **#38 (rmsnorm)** — the first sol-execbench → FlyDSL coverage target. This issue is the second coverage target: it extends coverage from elementwise+reduce to GEMM+epilogue.

## Why this issue exists

sol-execbench scores AI-generated kernel solutions across five NVIDIA-only
solution languages (PyTorch/Triton/CUTLASS/CuTe DSL/cuTile). For the AMD port
we need FlyDSL to be a scoreable solution language, and we need concrete
problems where a FlyDSL solution is wired into the eval harness, passes
correctness, and is benchmarked against the NVIDIA baseline.

The `jamba_attn_proj` problem is the right second target because:

- It is the simplest **GEMM-bearing** problem in the suite: a single matmul
  (`hidden_size=2560 × hidden_size=2560`) followed by a single elementwise
  residual add. No attention softmax, no MoE, no ragged/paged indexing.
- The reference is plain PyTorch (`torch.matmul(attn_output, o_proj_weight.t()) + residual`),
  so the correctness oracle is unambiguous.
- FlyDSL already ships production GEMM (`kernels/preshuffle_gemm.py`,
  `kernels/preshuffle_gemm_v2.py`) and an MFMA epilogue framework
  (`kernels/mfma_epilogues.py`). The work here is **not** "write a new GEMM
  from scratch" — it is "wire the existing FlyDSL GEMM + a residual-add
  epilogue into the sol-execbench solution contract, prove parity vs the
  PyTorch reference, and benchmark vs the cuTile baseline."
- The cuTile solution does **not** actually fuse — it launches `matmul_kernel`
  and `add_kernel` as two separate kernels with an intermediate global-memory
  write. The definition's reference comment describes the aspirational fusion
  ("compute matmul tiles in registers, add residual directly to register-held
  results, write final result to global memory"). The FlyDSL port should
  target the **fused** variant (residual add as a GEMM epilogue in registers)
  to demonstrate a win over the cuTile baseline on both correctness and perf.

**On the sibling cute_dsl variant (issue #39):** The `cute_dsl/jamba_attn_proj`
solution solves the identical problem but only kernels the residual add — the
matmul is left to `torch.matmul`. The cuTile variant kernels both the matmul
and the add. The FlyDSL port (fused GEMM+residual-add epilogue) is a **single
AMD solution** that covers both NVIDIA variants: it is a superset of the
cute_dsl variant (which only fuses the add) and a strict improvement over the
cuTile variant (which does not fuse at all). **Track both issues together**;
the FlyDSL kernel delivered here closes both. The cuTile variant (this issue)
is the **primary reference** because its NVIDIA solution is the more complete
one (both stages kernelled), making the cross-arch perf comparison
like-for-like.

## Goal (one sentence)

Implement, unit-test, and benchmark a FlyDSL fused GEMM+residual-add kernel
that satisfies the `ai21labs_AI21-Jamba-Reasoning-3B_attention_output_projection_with_residual`
definition from `amdpilot-org/SOL-ExecBench`, and report a clear pass/fail vs
the cuTile baseline on both MI300X (gfx942) and MI355X (gfx950).

## Hardware / workload

```text
Hardware:  1 × MI300X (gfx942) + 1 × MI355X (gfx950)
Models:    (none — this is a kernel benchmark, not an LLM workload)
Shape:     attn_output   [batch_size, seq_len, 2560] bfloat16
           residual      [batch_size, seq_len, 2560] bfloat16
           o_proj_weight [2560, 2560]                bfloat16
           output        [batch_size, seq_len, 2560] bfloat16
           (batch_size, seq_len) varies per workload.jsonl:
             (4, 128), (16, 128), (2, 293)
           GEMM effective: M = batch_size * seq_len, N = K = 2560
Baseline:  sol-execbench cuTile reference (examples/cutile/jamba_attn_proj/kernel.py)
           run on NVIDIA H100/B200 (cross-arch comparison)
           Fallback: PyTorch reference on AMD (torch.matmul + add)
Target:    max_abs_diff < 1e-2 vs PyTorch fp32 reference on all tested shapes
           kernel_ms < 0.9 × cuTile baseline on H100 (cross-arch) on ≥2 of 3
           workload shapes; OR kernel_ms < 0.9 × PyTorch torch.matmul on AMD
           on all 3 shapes (if cuTile-on-H100 numbers are unavailable)
```

The workload set (`workload.jsonl`) ships 3 concrete shapes. The cuTile
solution uses tile sizes `tm=128, tn=256, tk=64` for the GEMM and
`tm=32, tn=256` for the add kernel, with L2-swizzle (`GROUP_SIZE_M=8`) on
the GEMM grid. The parity/bench harness should cover the small
(`batch_size=4, seq_len=128` → M=512), mid (`batch_size=16, seq_len=128` →
M=2048), and odd (`batch_size=2, seq_len=293` → M=586, non-multiple of 128)
regimes at minimum.

## Scope

1. `kernels/sol_execbench/jamba_attn_proj.py` — FlyDSL kernel implementing
   `output = attn_output @ o_proj_weight.T + residual` as a **fused** GEMM
   with residual-add epilogue (residual loaded into registers and added to
   the accumulator before the global store, eliminating the intermediate
   `projected` tensor). Reuse the existing `kernels/preshuffle_gemm.py`
   builder and `kernels/mfma_epilogues.py` epilogue framework where possible
   — the new file is the sol-execbench-facing wrapper (entry point +
   destination-passing-style `run()`), not a GEMM reimplementation. If a
   fully-fused epilogue proves too risky for the first cut, a two-kernel
   fallback (GEMM + separate elementwise add, matching the cuTile solution
   structure) is acceptable for the correctness gate but must be flagged as
   a regression vs the fused target.
2. `tests/kernels/test_sol_execbench_jamba_attn_proj.py` — parity harness vs
   the fp32 PyTorch reference from `definition.json`. Tests all 3 workload
   shapes: `(4, 128)`, `(16, 128)`, `(2, 293)`. Random inputs at standard
   normal scale, `torch.manual_seed(0)` for determinism. The `(2, 293)`
   shape exercises the non-multiple-of-128 M dimension (M=586) — verify the
   kernel handles boundary tiles correctly (the cuTile solution uses
   `PaddingMode.ZERO` for this).
3. `tests/kernels/bench_sol_execbench_jamba_attn_proj.py` — reports
   `kernel_ms` per shape + comparison vs the cuTile baseline (run via the
   unified Docker image on H100/B200, same shape, same hardware) AND vs
   `torch.matmul` on AMD (as the cross-arch fallback baseline).
4. `assessment.md` (≤300 words) — ranking which shapes FlyDSL wins/ties/
   loses vs the cuTile baseline, with a one-paragraph hypothesis for any
   loss shapes (e.g. small-M is bandwidth-bound and the cuTile swizzle
   pattern is better matched to H100 L2; large-M is compute-bound and
   FlyDSL's MFMA preshuffle pipeline should win). Note whether the fused
   epilogue (single kernel) actually beats the two-kernel cuTile structure
   on AMD — if not, explain why (e.g. register pressure from holding the
   residual tile).

## Acceptance

- **Correctness:** `max_abs_diff < 1e-2` vs the PyTorch fp32 reference on
  all 3 workload shapes in `workload.jsonl`. This matches sol-execbench's
  default `max_atol=1e-2` from `docs/workload.md` (no per-workload tolerance
  override is set in this problem's `workload.jsonl`).
- **Performance:** `kernel_ms < 0.9 × cuTile_baseline` on ≥2 of the 3
  workload shapes, on at least one of MI300X / MI355X. If cuTile-on-H100
  numbers are unavailable, fall back to `kernel_ms < 0.9 × torch.matmul`
  on AMD on all 3 shapes.
- **Coverage:** works on both gfx942 (MI300X) and gfx950 (MI355X). The
  existing `preshuffle_gemm.py` already has arch-specific paths for both —
  the wrapper must exercise both.
- **Fusion:** the residual add is performed in registers as a GEMM epilogue
  (no intermediate global-memory write for `projected`). If a two-kernel
  fallback is shipped first, the fused variant must be landed as a follow-up
  before this issue is closed.

## Out of scope

- **The cute_dsl/jamba_attn_proj variant** — tracked in sibling issue #39.
  The FlyDSL kernel delivered here covers both variants (the cuTile variant
  is the superset). Do not file a separate FlyDSL kernel for the cute_dsl
  variant; close #39 as a duplicate of this issue once the kernel lands.
- **Backward pass / weight gradient** — sol-execbench's jamba_attn_proj
  problem is forward-only.
- **Other sol-execbench problems** — the rest of the suite (olmo3_post_norm,
  gqa_paged, moe_dispatch, etc.) are follow-ups. This issue is the
  **second** coverage target (after #38 rmsnorm), not a sweep.
- **Modifying the sol-execbench eval harness** — the wrapper conforms to
  the existing solution contract (`entry_point`, `destination_passing_style`,
  `dependencies`); no harness changes.
- **FP8 / INT8 quantized variants** — the definition specifies bfloat16
  only. FP8 GEMM is covered by the existing `kernels/preshuffle_gemm.py` FP8
  paths but is not in scope for this sol-execbench problem.

## Reproduction

To run the cuTile baseline locally (NVIDIA hardware required for cuTile):

```bash
# 1. Build the unified aiter+triton+flydsl Docker image (not yet built;
#    Dockerfile at /home/xiasun/unified-rocm.Dockerfile on the AMDPilot
#    fleet — ROCm 7.x base, Triton + AIter + FlyDSL installed side-by-side).
#    Note: cuTile (cuda.tile) requires NVIDIA CUDA toolchain; the cuTile
#    baseline must be run on an NVIDIA GPU (H100/B200), not on AMD.
docker build -f /home/xiasun/unified-rocm.Dockerfile -t unified-rocm:latest .

# 2. Clone the AMD fork and run the sol-execbench driver on the
#    jamba_attn_proj problem with the cuTile solution.
git clone https://github.com/amdpilot-org/SOL-ExecBench
cd SOL-ExecBench/examples/cutile/jamba_attn_proj
# (driver invocation TBD — depends on the sol-execbench runner CLI,
#  which is being wired up in parallel to this issue)
```

The FlyDSL solution is benchmarked by pointing the same driver at
`kernels/sol_execbench/jamba_attn_proj.py` instead of `kernel.py`, on AMD
hardware (MI300X / MI355X). For the cross-arch comparison, the cuTile
baseline numbers are captured on H100/B200 and recorded in `assessment.md`.

To run the PyTorch reference on AMD (fallback baseline):

```python
import torch
projected = torch.matmul(attn_output, o_proj_weight.t())
output = projected + residual
```

## Definition of done

- [ ] FlyDSL kernel implemented under `kernels/sol_execbench/jamba_attn_proj.py`
      (wrapping `kernels/preshuffle_gemm.py` + `kernels/mfma_epilogues.py`,
      fused residual-add epilogue, bfloat16).
- [ ] Parity test passes on MI300X (gfx942): `max_abs_diff < 1e-2` on all 3
      workload shapes `(4,128)`, `(16,128)`, `(2,293)`, deterministic on rerun.
- [ ] Parity test passes on MI355X (gfx950): same bound, exercising the
      gfx950-specific MFMA path.
- [ ] Benchmark shows `kernel_ms < 0.9 × cuTile_baseline` on ≥2 of 3
      workload shapes on at least one of MI300X / MI355X, OR `kernel_ms <
      0.9 × torch.matmul` on AMD on all 3 shapes.
- [ ] `assessment.md` (≤300 words) submitted with per-shape win/tie/loss
      table + hypothesis for losses + note on whether the fused epilogue
      beat the two-kernel cuTile structure on AMD.
- [ ] Sibling issue #39 (cute_dsl variant) closed as a duplicate of this
      issue, with a comment pointing to the landed FlyDSL kernel as the
      unified AMD solution for both NVIDIA variants.
- [ ] Issue closed with a one-line summary linking to the kernel, the test,
      the benchmark, and the assessment; names the next sol-execbench
      problem to port as a follow-up issue.

## ## Leaderboard baseline to beat

These are the current best public results on NVIDIA B200 (from the
[sol-execbench leaderboard](https://research.nvidia.com/benchmarks/sol-execbench/leaderboard))
for problem `010_attention_value_projection_with_transpose` (kernel_id=10). The agent's
FlyDSL kernel on MI300X/MI355X should aim to beat the **Scoring Baseline**
(optimized PyTorch) and approach the **top human/AI submissions**, normalized
by cross-arch roofline.

```text
GPU:           NVIDIA B200 (leaderboard) vs AMD MI300X/MI355X (this task)
Latency unit:  ms

SOL Bound (theoretical roofline):     0.010406 ms   (SOL-Score=1.0, speedup=2.88×)
Scoring Baseline (optimized PyTorch): 0.028237 ms   (SOL-Score=0.4544, speedup=0.92×)
Reference Implementation (plain PyTorch): 0.038863 ms   (SOL-Score=0.3368, speedup=0.67×)

Top public submissions:
  - **Rank 1** (`Recursive`): 0.024969 ms, SOL-Score=0.5101, speedup=1.04× vs reference
  - **Rank 2** (`ac4k`): 0.025401 ms, SOL-Score=0.5053
  - **Rank 3** (`Databricks`): 0.025716 ms, SOL-Score=0.5001
```

**Cross-arch note:** B200 timings are not directly comparable to MI300X/MI355X
(different GPU arch, different roofline). The agent should:
1. Measure the PyTorch reference on MI300X/MI355X (same shapes, same Docker image)
2. Compute the FlyDSL kernel's speedup vs the on-AMD PyTorch reference
3. Compare that speedup ratio to the leaderboard's `avg_speedup` ratios above
4. Target: FlyDSL speedup vs PyTorch-reference-on-AMD >= 0.8 × (Rank-1 speedup vs PyTorch-reference-on-B200)

## Definition of done

- [ ] FlyDSL kernel passes correctness (max_abs_diff < 1e-2 vs PyTorch reference on all tested workload shapes)
- [ ] FlyDSL kernel_ms measured on MI300X (gfx942) AND MI355X (gfx950) across >=5 workload shapes
- [ ] FlyDSL speedup vs PyTorch-reference-on-AMD >= 0.8 × Rank-1 speedup vs PyTorch-reference-on-B200 (=1.04× if top1 exists)
- [ ] `assessment.md` (<=300 words) ranks FlyDSL vs PyTorch eager on each shape, identifies win/tie/lose regimes
- [ ] Patch modifies actual kernel source (not just bench scripts) — bench-only patches are REJECTED

## Non-goals / rejection criteria

- Bench-only patches (no FlyDSL kernel source changes) are rejected
- Patches that set expected_value = measured_value (tautological pass) are rejected
- The metric contract must compare FlyDSL vs PyTorch reference on AMD, not an absolute ms
- "Completed" without a measured speedup ratio vs the on-AMD PyTorch reference is rejected

**Comment by sunxxuns** [comment]
## AMDPilot acceptance target

**The deliverable is a FlyDSL kernel.** A PyTorch reimplementation of this
problem -- with or without `torch.compile` -- is the *baseline*, not a result.
It does not close this issue no matter how closely it matches the reference.

### Implementation contract (hard requirement)

The submitted kernel MUST be written in FlyDSL and go through the FlyDSL
compiler. Concretely, the kernel module has to actually use the DSL, e.g.:

```python
import flydsl.compiler as flyc
from flydsl.expr import arith as ea, gpu, vector as ev
```

and dispatch through `flyc.jit` / `flyc.kernel`. A file placed under
`FlyDSL/kernels/sol_execbench/` that only calls `torch.*` ops does NOT satisfy
this issue, even if it is numerically perfect and even if it is faster.

If you conclude the port is not expressible in FlyDSL today, say so explicitly
and report exactly which DSL capability is missing. That is a legitimate,
useful outcome. Silently substituting a PyTorch implementation is not.

### Baseline definition

The baseline is the sol-execbench PyTorch reference (`reference.py`) running
eager, with `torch.compile` OFF. Measure it yourself in this run, on the same
GPU and the same workload shapes you benchmark your kernel on. Do not carry a
number over from another run, another GPU, or the issue text.

### Performance target

Your FlyDSL kernel must be **at least 10% faster than that measured baseline**
-- `kernel_ms <= 0.90 x baseline_ms` -- on **at least 80% of the tested
workload shapes**.

### Correctness target

`max_abs_diff < 1e-2` versus the PyTorch reference on **all** tested shapes. A
faster kernel that fails this is a failure, not a tradeoff.

### What you must report

Report a per-shape table containing, for every tested shape: the measured
eager baseline ms, your FlyDSL kernel ms, the ratio, and win/loss. Then state
the win rate against the 80% bar. A result without the baseline column next to
the kernel column is not checkable and will not be credited.
